src/physics/em/detail/UrbanMscScatter.hh

```diff
@@ -0,0 +1,542 @@
+            result.direction = rotate(from_spherical(cth, phi), inc_direction_);
+
+            // Displace the lateral position by the multiple scattering
+            if (input_.is_displaced && tau_ >= params_.tau_small)
+            {
+                real_type rmax2 = (true_path - geom_path)
+                                  * (true_path + geom_path);
+                if (rmax2 > 0)
+                {
+                    // Sample a unit direction of the displacement
+                    Real3 displacement = dir_displacement(rng, phi);
+
+                    // Rotate along the incident particle direction
+                    displacement = rotate(displacement, inc_direction_);
+
+                    // Scale with the lateral arm
+                    real_type arm = real_type(0.73) * std::sqrt(rmax2);
+                    for (auto i : range(3))
+                    {
+                        displacement[i] *= arm;
+                    }
+                    result.displacement = displacement;
+                }
+            }
+        }
+    }
+
+    return result;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Sample the scattering angle, \f$\theta\f$ at the end of the true step
+ * length, \f$t\f$, described in G4UrbanMscModel::SampleCosineTheta of the
+ * Geant4 10.7 release. See also, CERN-OPEN-2006-077 by L. Urban.
+ *
+ * The mean value of \f$u = \cos\theta\f$ follows \f$\exp(-t/\lambda_{1})\f$
+ * and the variance is written as \f$\frac{1+2e^{-\kappa r}}{3} - e^{-2r}\f$
+ * where \f$r = t/\lambda_{1}\f$ and \f$\kappa = \lambda_{1}/\lambda_{2}\f$.
+ * The \f$\cos\theta\f$ is sampled according to a model function of \f$u\f$,
+ * \f[
+ *   g(u) = q [ p g_1(u) + (1-p) g_2(u) ] - (1 - q) g_3(u)
+ * \f]
+ * where \f$p,q \in [0,1]\f$ and the functions \f$g_i\f$ have been chosen as
+ * \f$g_1(u) = c_1 e^{-a(1-u)}, g_2(u) = \frac{c_2}{(b-u)^d}, g_3(u) = c_3\f$
+ * with normalization constants, \f$d\f$. For small angles, \f$g_1\f$ is
+ * nearly Gaussian, \f$\exp(-\frac{\theta^{2}}{2\theta_{0}^{2}})\f$, if
+ * \f$\theta_0 \approx 1/1\f$, while \f$g_2\f$ has a Rutherford-like tail
+ * for large $\theta$, if \f$b \approx 1\f$ and \f$d\f$ is not far from 2.
+ */
+template<class Engine>
+CELER_FUNCTION real_type UrbanMscScatter::sample_cos_theta(Engine& rng,
+                                                           Energy  end_energy,
+                                                           real_type true_path,
+                                                           real_type limit_min)
+{
+    real_type cth     = 1.0;
+    real_type lambda  = helper_.msc_mfp(inc_energy_);
+    real_type lambda1 = helper_.msc_mfp(end_energy);
+
+    real_type tau = (std::abs(lambda1 - lambda) > lambda * 0.01 && lambda1 > 0)
+                        ? true_path * std::log(lambda / lambda1)
+                              / (lambda - lambda1)
+                        : true_path / lambda;
+
+    tau_ = tau;
```

Comment:
Also use `lambda_inc` and `lambda_end` instead of `{,1}`, `fabs` instead of `abs`, and/or maybe `soft_equal`:
```suggestion
    real_type tau = true_path * (lambda1 > 0 && !soft_equal(lambda, lambda1, 0.01))
                        ? std::log(lambda / lambda1)
                              / (lambda - lambda1)
                        : (1 / lambda);
```

Or maybe just make this a small helper function that takes the start/stop lambda as arguments, since they're not used elsewhere in the function. Then you could change the `lambda1 > 0` to a `CELER_EXPECT` because I don't think your mfp should ever be nonpositive?

Reply:
Okay,  minimal change for now with the suggested expression.